src/hooks/useQRCodeGenerator.jsx

```diff
@@ -1,78 +1,105 @@
-  //   const ...
+  const [showInput, setShowInput] = useState(true);
 
   // Function to generate a QR code from the input URL
-  const generateQRCode = () => {
+  const generateQRCode = async () => {
     // HINT 1: Utilize the qrccode library that converts a URL to a QR code data URL.
+  try {
+    const qrCodeDataUrl = await QRCode.toDataURL(url, {
+      width: 250,
+      margin: 5,
+      color: {
+      dark: '#141514',
+      light: '#FFFFFF'
+    }
+    });
+
     // Use the Import of the qrcode and chain to the native method toDataUrl() much like the example provided and specify the data within the object that you will be passing such as the {url, {width, margin, color:{dark, light}}} which containes the information to generate the qr-code and url. Lastly, this native method toDataUrl() will contain a callback function  that will update the qr variable and will also update the variable toggling the visibility of the input element.
-    QRCode.toDataURL(
-      console.log("Delete This Line OR Comment Out")
+
+
       // HINT 2: Ensure to pass the necessary parameters to the QR code generation method, such as the URL to convert and any styling options.
       // ...
       // HINT 3: Handle the callback of the QR code generation method, which provides the generated QR code data URL.
-      // ...
+      setQr(qrCodeDataUrl);
+      
       // HINT 4: Implement error handling to manage any issues that might occur during QR code generation.
-      // ...
+      setShowInput(false);
+        } catch (err) {
+          console.error(err);
+        }
       // HINT 5: Update the relevant state variables with the generated QR code data URL and adjust the UI accordingly.
       // ...
       // HINT 6: Consider the user experience and how the UI should change once the QR code has been generated.
       // ...
-    );
+    ;
+  
   };
-
+  
   // Function to download the generated QR code as a PNG file
   const downloadQRCode = () => {
     // HINT 1: Consider encapsulating the filename prompting logic into a separate function.
     const getFileName = () => {
       // HINT 2: Use a method to prompt the user for input and store the response.
-      // ...
+      let filename = prompt('Please enter file name', 'QRcode')
       // HINT 3: Implement a check for an empty filename and utilize recursion to re-prompt the user if necessary.
-      // ...
-      // HINT 4: Ensure the function returns the obtained filename.
-      // ...
+      if (!filename) {
+        filename = getFileName();
+      }
+      // HINT 4 & 6: Ensure the function returns the obtained filename.
+      return filename.replace(/[^a-z0-9]/gi, '_').toLowerCase();
     };
 
     // HINT 5: Call the above function to retrieve a filename and store it in a variable.
-    // ...
-
-    // HINT 6: Format the filename to ensure it is filesystem-friendly.
-    // ...
+    const filename = `${getFileName()}.png`;
```

Comment:
Well done. The download works well and the name is corrected when needed. The cancel button is not wokring - but I did not test that for myself either, so might be the same!